lib/Transforms/MemoryBanking.cpp

```diff
@@ -324,43 +320,111 @@ void updateFuncOpArgAttrs(func::FuncOp funcOp, unsigned argIndex,
   for (unsigned i = 0; i < numInsertedArgs; ++i) {
-    updatedArgAttrs[argIndex + 1 + i] =
-        DictionaryAttr::get(funcOp.getContext());
+    updatedArgAttrs[argIndex + 1 + i] = remainingAttrs;
   }
 
   // Set the updated attributes.
   funcOp->setAttr("arg_attrs",
                   ArrayAttr::get(funcOp.getContext(), updatedArgAttrs));
 }
 
-SmallVector<Value, 4> createBanks(Value originalMem, unsigned bankingFactor,
-                                  std::optional<int> bankingDimensionOpt) {
+unsigned getCurrBankingInfo(BankingConfigAttributes bankingConfigAttrs,
+                            StringRef attrName) {
+  auto getFirstInteger = [](Attribute attr) -> unsigned {
+    if (auto arrayAttr = dyn_cast<ArrayAttr>(attr)) {
+      assert(!arrayAttr.empty() &&
+             "BankingConfig ArrayAttr should not be empty");
+      auto intAttr = dyn_cast<IntegerAttr>(arrayAttr.getValue().front());
+      assert(intAttr && "BankingConfig elements must be integers");
+      return intAttr.getInt();
+    }
+    auto intAttr = dyn_cast<IntegerAttr>(attr);
+    assert(intAttr && "BankingConfig attribute must be an integer");
+    return intAttr.getInt();
+  };
+
+  if (attrName.str() == bankingFactorsStr) {
+    return getFirstInteger(bankingConfigAttrs.factors);
+  }
+
+  assert(attrName.str() == bankingDimensionsStr &&
+         "BankingConfig only contains 'factors' and 'dimensions' attributes");
+  return getFirstInteger(bankingConfigAttrs.dimensions);
+}
+
+Attribute getRemainingBankingInfo(MLIRContext *context,
+                                  BankingConfigAttributes bankingConfigAttrs,
+                                  StringRef attrName) {
+  if (attrName.str() == bankingFactorsStr) {
+    if (auto arrayAttr = dyn_cast<ArrayAttr>(bankingConfigAttrs.factors)) {
+      assert(!arrayAttr.empty() &&
+             "BankingConfig factors ArrayAttr should not be empty");
+      return arrayAttr.size() > 1
+                 ? ArrayAttr::get(context, arrayAttr.getValue().take_back(
+                                               arrayAttr.size() - 1))
+                 : nullptr;
+    }
+    auto intAttr = dyn_cast<IntegerAttr>(bankingConfigAttrs.factors);
+    assert(intAttr && "BankingConfig factor must be an integer");
+    return nullptr;
+  }
+
+  assert(attrName.str() == bankingDimensionsStr &&
+         "BankingConfig only contains 'factors' and 'dimensions' attributes");
+  if (auto arrayAttr = dyn_cast<ArrayAttr>(bankingConfigAttrs.dimensions)) {
+    assert(!arrayAttr.empty() &&
+           "BankingConfig dimensions ArrayAttr should not be empty");
+    return arrayAttr.size() > 1
+               ? ArrayAttr::get(context, arrayAttr.getValue().take_back(
+                                             arrayAttr.size() - 1))
+               : nullptr;
+  }
+  auto intAttr = dyn_cast<IntegerAttr>(bankingConfigAttrs.dimensions);
+  assert(intAttr && "BankingConfig dimension must be an integer");
+  return nullptr;
```

Comment:
Can we refactor this too? It looks identical to me.

Reply:
sure will do!